editor/src/components/canvas/remix/remix-utils.tsx

```diff
@@ -99,3 +128,397 @@ export function getRoutesFromRouteManifest(
+  topLevelComponentRendererComponents: React.MutableRefObject<
+    MapLike<MapLike<ComponentRendererComponent>>
+  >,
+  fileBlobs: CanvasBase64Blobs,
+  hiddenInstances: Array<ElementPath>,
+  displayNoneInstances: Array<ElementPath>,
+): {
+  executionScopeCreator: (innerProjectContents: ProjectContentTreeRoot) => ExecutionScope
+  loader: LoaderFunction | undefined
+  action: ActionFunction | undefined
+  rootComponentUid: string
+} {
+  const executionScopeCreator = (innerProjectContents: ProjectContentTreeRoot) => {
+    let resolvedFiles: MapLike<Array<string>> = {}
+    let resolvedFileNames: Array<string> = ['/src/root.js']
+
+    const requireFn = curriedRequireFn(innerProjectContents)
+    const resolve = curriedResolveFn(innerProjectContents)
+
+    const customRequire = (importOrigin: string, toImport: string) => {
+      if (resolvedFiles[importOrigin] == null) {
+        resolvedFiles[importOrigin] = []
+      }
+      let resolvedFromThisOrigin = resolvedFiles[importOrigin]
+
+      const alreadyResolved = resolvedFromThisOrigin.includes(toImport) // We're inside a cyclic dependency, so trigger the below fallback     const filePathResolveResult = alreadyResolved
+        ? left<string, string>('Already resolved')
+        : resolve(importOrigin, toImport)
+
+      forEachRight(alreadyResolved, (filepath) => resolvedFileNames.push(filepath))
+
+      const resolvedParseSuccess: Either<string, MapLike<any>> = attemptToResolveParsedComponents(
+        resolvedFromThisOrigin,
+        toImport,
+        innerProjectContents,
+        customRequire,
+        mutableContextRef,
+        topLevelComponentRendererComponents,
+        '/src/root.js',
+        {},
+        [],
+        [],
+        metadataContext,
+        NO_OP,
+        false,
+        alreadyResolved,
+        null,
+      )
+      return foldEither(
+        () => {
+          // We did not find a ParseSuccess, fallback to standard require Fn
+          return requireFn(importOrigin, toImport, false)
+        },
+        (scope) => {
+          // Return an artificial exports object that contains our ComponentRendererComponents
+          return scope
+        },
+        resolvedParseSuccess,
+      )
+    }
+    return createExecutionScope(
+      filename,
+      customRequire,
+      mutableContextRef,
+      topLevelComponentRendererComponents,
+      innerProjectContents,
+      filename,
+      fileBlobs,
+      hiddenInstances,
+      displayNoneInstances,
+      metadataContext,
+      NO_OP,
+      false,
+      null,
+    )
+  }
+
+  const executionScope = executionScopeCreator(projectContents)
+
+  const nameAndUid = getDefaultExportNameAndUidFromFile(projectContents, filename)
+
+  return {
+    executionScopeCreator: executionScopeCreator,
+    loader: executionScope.scope['loader'] as LoaderFunction | undefined,
+    action: executionScope.scope['action'] as ActionFunction | undefined,
```

Comment:
I think we should attach a `FIXME` to these, because really we should be creating the `executionScope` at the point where we use the `loader` and `action` like we do for the module's default export component

Reply:
added the FIXME comment in https://github.com/concrete-utopia/utopia/pull/4162/commits/cddc103ffd7070810f4ed2a17a35d4afc97a1f4d